crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,361 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
+    } else {
+      // Finish AAD by applying padding
+      if (cipher_ctx->pad_aad) {
+        remainder = cipher_ctx->len.aad % POLY1305_TAG_LEN;
+        if (remainder != 0) {
+          static const uint8_t padding[POLY1305_TAG_LEN] = {0};
+          CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+        }
+        cipher_ctx->pad_aad = 0;
+      }
+      // cipher/plain text |EVP_CipherUpdate|
+      if (EVP_CIPHER_CTX_encrypting(ctx)) {
+        // Encryption
+        cipher_chacha20_do_cipher(ctx, out, inp, inl);
+        // Update poly1305 with computed ciphertext
+        CRYPTO_poly1305_update(poly_ctx, out, inl);
+        cipher_ctx->len.text += inl;
+      } else {
+        // Decryption
+        // Update poly1305 with incoming ciphertext
+        CRYPTO_poly1305_update(poly_ctx, inp, inl);
+        cipher_chacha20_do_cipher(ctx, out, inp, inl);
+        cipher_ctx->len.text += inl;
+      }
+    }
+  }
+
+  // Process an |EVP_CipherFinal|
+  if (inp == NULL) {
+    uint8_t temp[POLY1305_TAG_LEN];
+    static const uint8_t padding[POLY1305_TAG_LEN] = {0};
+
+    // Finish AAD inp case there were no intermediate Update() calls
+    if (cipher_ctx->pad_aad) {
+      remainder = cipher_ctx->len.aad % POLY1305_TAG_LEN;
+      if (remainder != 0) {
+        CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+      }
+      cipher_ctx->pad_aad = 0;
+    }
+
+    // Apply padding for the text
+    remainder = cipher_ctx->len.text % POLY1305_TAG_LEN;
+    if (remainder != 0) {
+      CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+    }
+
+    // ChaCha20-Poly1305 passes the AAD and CT lengths through Poly1305 as two
+    // 64-bit little-endian integers.
+#ifdef OPENSSL_BIG_ENDIAN
+    uint8_t length_bytes[2 * sizeof(uint64_t)];
+    CRYPTO_store_u64_le(length_bytes, cipher_ctx->len.aad);
+    CRYPTO_store_u64_le(length_bytes + sizeof(uint64_t), cipher_ctx->len.text);
+#else
+    // For a little-endian platform, the struct's layout in memory works as-is.
+    const uint8_t *length_bytes = (const uint8_t *) &cipher_ctx->len;
+#endif
+    CRYPTO_poly1305_update(poly_ctx, length_bytes, 2 * sizeof(uint64_t));
+
+    // Compute the tag and write it to scratch or the cipher context
+    CRYPTO_poly1305_finish(poly_ctx, EVP_CIPHER_CTX_encrypting(ctx) ?
+      cipher_ctx->tag : temp);
+    cipher_ctx->poly_initialized = 0;
+
+    // Check the tags if we're decrypting
+    if (!EVP_CIPHER_CTX_encrypting(ctx)) {
+      if (CRYPTO_memcmp(temp, cipher_ctx->tag, cipher_ctx->tag_len)) {
+        return -1;
+      }
+    }
+  }
+  return (int32_t) inl;
+}
+
+static void cipher_chacha20_poly1305_cleanup(EVP_CIPHER_CTX *ctx) {
+  if (ctx->cipher_data) {
+    OPENSSL_cleanse(ctx->cipher_data, sizeof(CIPHER_CHACHA_POLY_CTX));
+  }
+}
+
+// Returns a pointer to the first 64-byte alignment within |poly_ctx|.
+static void *cipher_chacha20_poly1305_aligned_poly(EVP_CIPHER_CTX *ctx) {
```

Comment:
Should we make this function the same as https://github.com/aws/aws-lc/blob/7e6aef83ecf7bec3a0c6d38e38c64ac079647c78/crypto/poly1305/poly1305.c#L45?